modules/permission.js

```diff
@@ -1,38 +1,195 @@
+const Offering = models.Offering;
+const Role = models.Role;
+const User = models.User;
+const UserOffering = models.UserOffering;
 
+async function updateCoursePermission(userId, courseOfferingId, roleName) {
+  try {
+      var role = await Role.findOrCreate({
+        where: { roleName: roleName },
+        defaults: { id: uuid() }
+      });
+      
+      var roleInfo = role[0].dataValues;
+      var userOfferingResult = await UserOffering.findOrCreate({
+        where: { userId: userId, courseOfferingId: courseOfferingId },
+        defaults: { roleId: roleInfo.id }});
+      
+      if (userOfferingResult[1] === true) { 
+        return true; 
+      } else {
+        
+        var updateResult = await UserOffering.update(
+          { roleId: roleInfo.id },
+          {
+            where: {
+              userId: userId,
+              courseOfferingId: courseOfferingId
+            }
+          });
 
-// const client = require('./../../modules/redis');
-const macl = require('acl');
-const aclSeq = require('acl-sequelize');
-const db = require('../models/index');
-const acl = new macl(new aclSeq(db.sequelize, { prefix : 'acl_'}));
+        return null !== updateResult;
+      }
+    } catch (err) {
+      console.log(err);
+      return false;
+    }
+}
+
+async function addCoursePermission(userId, courseOfferingId, permission) {
+  switch (permission) {
+    case 'View': return await updateCoursePermission(userId, courseOfferingId, 'Student');
+    case 'Edit': return await updateCoursePermission(userId, courseOfferingId, 'Instructor');
+    case 'Delete': return await updateCoursePermission(userId, courseOfferingId, 'Administrator');
+    default: return false; 
+  }
+}
+
+async function removeCoursePermission(userId, courseOfferingId) {
+  switch (permission) {
+    case 'Delete': return await updateCoursePermission(userId, courseOfferingId, 'Instructor');
+    case 'Edit': return await updateCoursePermission(userId, courseOfferingId, 'Student');
+    case 'View': return await UserOffering.destroy( {where : {userId : userId, courseOfferingId : courseOfferingId}});
+    default: return false;
+  }
+}
+
+async function isWatchingUniversityAllowed(userData, courseData) {
+  if (userData === null || courseData === null) { return false; }
+
+  var courseUniversity = await Offering.findOne({where : {id : courseData.offeringId}});
 
-// acl = new macl(new macl.redisBackend(client, "ClassTranscribe::acl::"));
-// var client_api = require('../../db/db');
+  if (courseUniversity === null) return false;
 
-function addUser(userID) {
-  acl.addUserRoles(userID, userID);
+  courseUniversity = courseUniversity.dataValues;
+
+  return userData.universityId === courseUniversity.universityId;
 }
 
-function addCoursePermission(userID, classID, permission) {
-  acl.allow(userID, classID, permission);
+async function isWatchingPrivateAllowed(userData, courseData) {
+  if (userData === null || courseData === null) { return false; }
```

Comment:
I might not have understood this right, but.
If a user has access to course A, but not course B. This function would allow him to watch course B nonetheless, right?

Reply:
`isWatchingPricateAllowed` takes one courseOfferingId as input. Same as `isManagingAllowed`.

Reply:
True, but where are you using the input?

Reply:
Ah, forgot to add the courseOfferingId constraint. My bad.